packages/sdk-js/src/sticky-bucket-service.ts

```diff
@@ -0,0 +1,274 @@
+  abstract getAssignments(
+    attributeName: string,
+    attributeValue: string
+  ): Promise<StickyAssignmentsDocument | null>;
+
+  abstract saveAssignments(doc: StickyAssignmentsDocument): Promise<unknown>;
+
+  /**
+   * The SDK calls getAllAssignments to populate sticky buckets. This in turn will
+   * typically loop through individual getAssignments calls. However, some StickyBucketService
+   * instances (i.e. Redis) will instead perform a multi-query inside getAllAssignments instead.
+   */
+  async getAllAssignments(
+    attributes: Record<string, string>
+  ): Promise<Record<StickyAttributeKey, StickyAssignmentsDocument>> {
+    const docs: Record<string, StickyAssignmentsDocument> = {};
+    (
+      await Promise.all(
+        Object.entries(attributes).map(([attributeName, attributeValue]) =>
+          this.getAssignments(attributeName, attributeValue)
+        )
+      )
+    ).forEach((doc) => {
+      if (doc) {
+        const key = `${doc.attributeName}||${doc.attributeValue}`;
+        docs[key] = doc;
+      }
+    });
+    return docs;
+  }
+}
+
+export class LocalStorageStickyBucketService extends StickyBucketService {
+  private prefix: string;
+  private localStorage: LocalStorageCompat | undefined;
+  constructor({
+    prefix = "gbStickyBuckets::",
+    localStorage,
+  }: {
+    prefix?: string;
+    localStorage?: LocalStorageCompat;
+  } = {}) {
+    super();
+    this.prefix = prefix;
+    this.localStorage = localStorage;
+    try {
+      if (!this.localStorage && globalThis.localStorage) {
+        this.localStorage = globalThis.localStorage;
+      }
+    } catch (e) {
+      // Ignore localStorage errors
+    }
+  }
+  async getAssignments(attributeName: string, attributeValue: string) {
+    const key = `${attributeName}||${attributeValue}`;
+    let doc: StickyAssignmentsDocument | null = null;
+    if (!this.localStorage) return doc;
+    try {
+      const raw = await this.localStorage.getItem(this.prefix + key);
+      const data = JSON.parse(raw || "{}");
+      if (data.attributeName && data.attributeValue && data.assignments) {
+        doc = data;
+      }
+    } catch (e) {
+      // Ignore localStorage errors
+    }
+    return doc;
+  }
+  async saveAssignments(doc: StickyAssignmentsDocument) {
+    const key = `${doc.attributeName}||${doc.attributeValue}`;
+    if (!this.localStorage) return;
+    await this.localStorage.setItem(this.prefix + key, JSON.stringify(doc));
+  }
+}
+
+export class ExpressCookieStickyBucketService extends StickyBucketService {
+  /** intended to be used with cookieParser() middleware from npm: 'cookie-parser' **/
+  private prefix: string;
+  private req: RequestCompat | undefined;
+  private res: ResponseCompat | undefined;
+  private cookieAttributes: CookieAttributes;
+  constructor({
+    prefix = "gbStickyBuckets::",
+    req,
+    res,
+    cookieAttributes = {},
+  }: {
+    prefix?: string;
+    req?: RequestCompat;
+    res?: ResponseCompat;
```

Comment:
For all of these, the arguments should be required so it can be caught easier at compile-time.  In this example, if `req` is not provided, it will throw an exception, but Typescript or eslint won't catch the bug.